src/train_cmd.cpp

```diff
@@ -985,7 +986,7 @@ static CommandCost CheckTrainAttachment(Train *t)
 		t->SetNext(nullptr);
 
 		/* Don't check callback for articulated or rear dual headed parts */
-		if (!t->IsArticulatedPart() && !t->IsRearDualheaded()) {
+		if (t->IsEngine() && !t->IsArticulatedPart() && !t->IsRearDualheaded()) {
```

Comment:
This is supposed to check whether this is a real train. I.e. IsEngine should check the first vehicle.
However, "t" can be about anything here, and you do not want to call the attach-callback for free wagon chains.